include/zephyr/drivers/counter.h

```diff
@@ -189,6 +189,7 @@ struct counter_top_cfg {
 /** @brief Structure with generic counter features.
  */
 struct counter_config_info {
+	DEVICE_MMIO_ROM;
```

Comment:
Is there a reason we couldn't add the `DEVICE_MMIO_ROM` declaration to the `config` structure in the GPT counter, rather than changing common code here? I'd like to avoid touching common headers here if we can

Reply:
Currently some API definitions in counter.h is assuming that dev->config is `counter_config_info`,  for example:
`static inline bool z_impl_counter_is_counting_up(const struct device *dev)
{
	const struct counter_config_info *config =
			(const struct counter_config_info *)dev->config;

	return config->flags & COUNTER_CONFIG_INFO_COUNT_UP;
}`
So we need to use `counter_config_info`  as `config` in DEVICE_DT_INST_DEFINE(), then we also need to put `DEVICE_MMIO_ROM` as the first field of `counter_config_info`

Reply:
Well this is fun :). So this is a generic issue that would appear for any MMU system using the counter API? Could we leverage `DEVICE_MMIO_NAMED_xxx` macros to avoid the requirement that the MMIO structure be the first element?

Reply:
Now I changed to use `DEVICE_MMIO_NAMED_xxx`, thanks @danieldegrasse for your suggestions.